x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -690,6 +694,45 @@ void writeAuthToContext(Authentication authentication) {
         private void authenticateToken(AuthenticationToken token) {
             this.consumeToken(token);
         }
+
+        private User maybeMergeAnonymousRolesForUser(User user) {
+            if (SystemUser.is(user) || XPackUser.is(user) || XPackSecurityUser.is(user) || AsyncSearchUser.is(user)) {
+                return user;
+            } else if (isAnonymousUserEnabled && anonymousUser.equals(user) == false) {
+                if (anonymousUser.roles().length == 0) {
+                    throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+                }
+                User userWithMergedRoles = new User(
+                    user.principal(),
+                    mergeAnonymousRoles(user.roles()),
+                    user.fullName(),
+                    user.email(),
+                    user.metadata(),
+                    user.enabled()
+                );
+                if (user.isRunAs()) {
+                    final User authenticatedUserWithMergedRoles = new User(
+                        user.authenticatedUser().principal(),
+                        mergeAnonymousRoles(user.authenticatedUser().roles()),
+                        user.authenticatedUser().fullName(),
+                        user.authenticatedUser().email(),
+                        user.authenticatedUser().metadata(),
+                        user.authenticatedUser().enabled()
+                    );
+                    userWithMergedRoles = new User(userWithMergedRoles, authenticatedUserWithMergedRoles);
+                }
+                return userWithMergedRoles;
+            } else {
+                return user;
+            }
+        }
+
+        private String[] mergeAnonymousRoles(String[] existingRoles) {
```

Comment:
nit: this implies that we have some special handling when merging the anonymous roles, while we just merge the two String arrays. I get that this is cleaner than doing it twice above but maybe a more generic `mergeRoles` where you pass both existingRoles and anonymoysUser.roles? Just a suggestion though

Reply:
This is a good suggestion. It is cleaner that way. I [updated](https://github.com/elastic/elasticsearch/pull/53453/files#diff-1436647fa05b4a2d5b2712fdefafa98bR730).